File: src/frontend/rate-my-fit/src/API/item.ts
import { APICall } from './API';
import { Form } from '../types/item';

export const addItem = async (itemData: Form) => {
    const response = await APICall({
        method: 'POST',
        path: '/item/add',
        headers: null,
        body: itemData,
        multipart: false
    });

    return response;
};

export const getItem = async (itemID: string) => {
    const response = await APICall({
        method: 'GET',
        path: `/item/${itemID}`,
        headers: null,
        body: null,
        multipart: false
    });

    return response;
};

export const updateItem = () => {

};

export const getBrandItems = () => {

};

export const getAllBrands = async () => {
    const response = await APICall({
        method: 'GET',
        path: '/item/all/brands',
        headers: null,
        body: null,
        multipart: false
    });

    return response;
};

export const searchItemName = async (itemName: string, limit: number) => {
    const response = await APICall({
        method: 'GET',
        path: `/item/search/name?itemName=${itemName}&limit=${limit}`,
        headers: null,
        body: null,
        multipart: false
    });

    return response;
};

File: src/backend/routes/item.py
from fastapi import APIRouter, Depends

from dao.aggregator import DAO
from models.item import ItemModel
from services.item import ItemService


router = APIRouter()


def get_service() -> ItemService:
    return ItemService(DAO())


@router.post("/add", response_model=None)
async def add_item(
    item_data: ItemModel, service: ItemService = Depends(get_service)
) -> dict:
    item_data = item_data.model_dump()
    result = service.add(item_data)

    return {"message": "Successful adding", "data": result}


@router.get("/{item_id}", response_model=None)
async def read_item(item_id: str, service: ItemService = Depends(get_service)) -> dict:
    return {"message": "Successful retrieving", "data": service.get(item_id)}


@router.put("/{item_id}", response_model=None)
async def update_item(
    item_id: str, item_data: ItemModel, service: ItemService = Depends(get_service)
) -> dict:
    item_data = item_data.model_dump()
    result = service.update(item_id, item_data)

    return {"message": "Successful update", "data": result}


@router.get("/by/{brand_name}", response_model=None)
async def get_brand_items(
    brand_name: str,
    skip: int = None,
    limit: int = None,
    service: ItemService = Depends(get_service),
) -> list[dict]:
    return {
        "message": "Successful retrieving",
        "data": service.get_by_brand(brand_name, skip, limit),
    }


@router.get("/all/brands", response_model=None)
async def get_all_brands(service: ItemService = Depends(get_service)) -> dict:
    return {"message": "Successful retrieving", "data": service.get_brands()}


@router.get("/search", response_model=None)
async def search_item(
    item_name: str, limit: int = 15, service: ItemService = Depends(get_service)
) -> dict:
    return {
        "message": "Successful retrieving",
        "data": service.search(item_name, limit),
    }
